src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export enum PrincipalType {
+  Users = 'users',
+  Groups = 'groups',
+}
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export type Permissions = Record<string, Principals>;
+
+export interface TransformedPermission {
+  type: string;
+  name: string;
+  permissions: string[];
+}
+
+const addToPrincipals = ({
+  principals = {},
+  users,
+  groups,
+}: {
+  principals: Principals;
+  users?: string[];
+  groups?: string[];
+}) => {
+  if (!!users) {
+    if (!principals.users) {
+      principals.users = [];
+    }
+    principals.users = Array.from(new Set([...principals.users, ...users]));
+  }
+  if (!!groups) {
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deleteFromPrincipals = ({
+  principals,
+  users,
+  groups,
+}: {
+  principals?: Principals;
+  users?: string[];
+  groups?: string[];
+}) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+const checkPermission = (currentPrincipals: Principals | undefined, principals: Principals) => {
+  return (
+    (currentPrincipals?.users &&
+      principals?.users &&
+      checkPermissionForSinglePrincipalType(currentPrincipals.users, principals.users)) ||
+    (currentPrincipals?.groups &&
+      principals?.groups &&
+      checkPermissionForSinglePrincipalType(currentPrincipals.groups, principals.groups))
+  );
+};
+
+const checkPermissionForSinglePrincipalType = (
+  currentPrincipalArray: string[],
+  principalArray: string[]
+) => {
+  return (
+    currentPrincipalArray &&
+    principalArray &&
+    (currentPrincipalArray.includes('*') ||
+      principalArray.some((item) => currentPrincipalArray.includes(item)))
+  );
+};
+
+export class ACL {
+  private permissions?: Permissions;
+  constructor(initialPermissions?: Permissions) {
+    this.permissions = initialPermissions || {};
+  }
+
```

Comment:
ACL and Permissions seems the the same thing, shall we consider to move the functions to Permissions to reduce the complexity?

Reply:
You mean change the class name `ACL` to `Permissions`? `Permissions` is a permissionType->Principals map, just a object which has the same structure with the field in .kibana index, ACL contains the Permissions object and some fundamental functions, I think they are not the same.

Reply:
by simply looking at the code, the `ACL` class has only one field, which is `permissions?: Permissions` , so all operations `ACL` can provide is based on the `permissions` right? so just don't feel we don't need 2 levels of classes here

Reply:
`export type Permissions = Record<string, Principals>;`

`Permissions` is a type, not class, and I think we need two level of definitions here is because `Permissions` is referenced by saved_object_client and repository, the relationship between `ACL` and `Permissions` just like the relationship between [SavedObjectsSerializer](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/9e3e3a7ae8e950c5e4137ebf5dafd8f093099f0b/src/core/server/saved_objects/serialization/serializer.ts#L44) and `SavedObjectsRawDoc`.

Reply:
okay, the name `ACL` doesn't sounds right here, maybe `PermissionEvaluator` makes more sense

Reply:
And should this go here in the saved object service? or workspace service?

Reply:
Since the core module in OSD cannot depend on the code of plugin, but we have to add the field `permissions` to `SavedObjectsRepository`, `saved_objects_client` and `saved_objects` in the core module, so we cannot move the ACL class to the workspace plugin.